main.go

```diff
@@ -0,0 +1,276 @@
+		Envar("HIERARCHY_DEBUG").Default("false").BoolVar(&cfg.logDebug)
+
+	_, err := application.Parse(os.Args[1:])
+
+	if cfg.printVersion {
+		version.Print()
+		os.Exit(0)
+	}
+
+	if err != nil {
+		fmt.Fprintln(os.Stderr, errors.Wrapf(err, "Error parsing commandline arguments"))
+		application.Usage(os.Args[1:])
+		os.Exit(2)
+	}
+	return cfg
+}
+
+// checkForError fails the program with a fatal error message if e != nil
+func checkForError(e error) {
+	if e != nil {
+		log.Fatal(e)
+	}
+}
+
+func main() {
+	cfg := parseFlags()
+
+	// Get the logging configured the way we want it
+	log.SetOutput(os.Stdout)
+	if cfg.logTrace {
+		log.SetLevel(log.TraceLevel)
+	} else if cfg.logDebug {
+		log.SetLevel(log.DebugLevel)
+	} else {
+		log.SetLevel(log.InfoLevel)
+	}
+
+	version.Log()
+
+	cfg.basePath = path.Dir(cfg.hierarchyFile)
+
+	log.WithFields(log.Fields{
+		"hierarchyFile":     cfg.hierarchyFile,
+		"basePath":          cfg.basePath,
+		"outputFile":        cfg.outputFile,
+		"outputPermissions": cfg.outputFile,
+		"filterExtension":   cfg.filterExtension,
+		"failMissing":       cfg.failMissing,
+	}).Debug("Configuration settings")
+
+	// Make sure we remove the output file if it already exists
+	// Just in case the program ends for any reason other than success
+	// We don't want to give the impression that we completed the merging
+	if _, err := os.Stat(cfg.outputFile); err == nil {
+		log.WithFields(log.Fields{
+			"path": cfg.outputFile,
+		}).Info("Removing existing output file")
+		err := os.Remove(cfg.outputFile)
+		checkForError(err)
+	}
+
+	// process the hierarchy and get the list of include files
+	hierarchy := processHierarchy(cfg)
+
+	// Lets do the deed
+	mergeYamls(hierarchy, cfg.filterExtension, cfg.outputFile)
+}
+
+// processHierarchy loads the hiearchy file and generates a list
+// of folders to be processed
+func processHierarchy(cfg config) []string {
+	hierarchy := []string{}
+	hierarchyFile, err := os.Open(cfg.hierarchyFile)
+	checkForError(err)
+	defer hierarchyFile.Close()
+
+	// Start reading from the file with a reader.
+	reader := bufio.NewReader(hierarchyFile)
+	var line string
+	for {
+		line, err = reader.ReadString('\n')
+		if err != nil && err != io.EOF {
+			break
+		}
+
+		// Trim spaces and comments
+		includePath := strings.Split(line, "#")[0]
+		includePath = strings.TrimSpace(includePath)
+		// TODO write own ExpandEnv that fails if variable is not defined
+		// TODO replace environment variables expansion with go template
+		// Example: https://gist.github.com/raphink/e1afb23300a7e53a0949
+		includePath = os.ExpandEnv(includePath)
+		// Process path
+		if len(includePath) > 0 {
+			includePath = path.Join(cfg.basePath, includePath)
+			// Check if directory exists
+			if stat, err := os.Stat(includePath); err == nil && stat.IsDir() {
+				hierarchy = append(hierarchy, includePath)
+				log.WithFields(log.Fields{
+					"path": includePath,
+				}).Debug("Adding path to hierarchy")
```

Comment:
Would it be beneficial to add an absolute path for debugging as well, like below? What do you think?

```suggestion
				hierarchy = append(hierarchy, includePath)
				absPath, _ := filepath.Abs(includePath)
				log.WithFields(log.Fields{
					"path": includePath,
					"abs_path": absPath,
				}).Debug("Adding path to hierarchy")
```